fix datapreprocessing clonability

franchuterivera opened the pull request:
While debugging #921  I found out that column transformer is not an active attribute of the class dict, which means that if clone is used, a fitted datapreprocesor will be cloned as if it was not fit.

Why?
https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/base.py#L85
Sklearn clone does a deep copy only on the attributes of the class defined in the init.

If we have this attribute, the error message seen in 921 will be better (I also made this more understandable to the user). 

Nevertheless, #921 problem was something else (see issue for details), this prevent problems in the future.

mfeurer commented:
Hey, could you briefly explain why a fitted preprocessor gets cloned in the first place? The fact that we have to work around sklearn's clone function worries me quite a bit so I need to understand what's going on here.

franchuterivera commented:
My comment is to comply with https://github.com/automl/auto-sklearn/issues/876

We are not cloning estimators, so it should not affect. But, to comply with scikit learn API, we need to make sure that a estimator has all their parameters available via get/set parameters, that is, in the __init__ method of the class

franchuterivera commented:
Problem statement:

Autosklearn relies on the check is fitted method to decide whether or not to use cv models or models to predict.

This can be seen in:
https://github.com/automl/auto-sklearn/blob/859fb993607bd5cfb881aca3b346cf4d1addf470/autosklearn/automl.py#L599


There is a problem with the check is fitted method from sklearn. This method uses any variable having underscore at the end, as a criteria to see if a model is fitted. Please see:
https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/utils/validation.py#L1015

I added a check to enforce we comply with this. I also added a CV check in regression because I was not able to find a method that makes sure CV works in the case of regression.

And lastly, I found I bug in the testing that instead of combinations what I really intended to use was product() in itertools, and this really get all possible combinations.

mfeurer commented on test/test_pipeline/test_classification.py:
```diff
@@ -822,3 +823,37 @@ def test_fit_instantiates_component(self):
+            for name, step in auto.named_steps.items():
+                with self.assertRaisesRegex(sklearn.exceptions.NotFittedError,
+                                            "instance is not fitted yet"):
+                    check_is_fitted(step)
```
Would it make sense to also call this on the full pipeline? So `check_is_fitted(auto)`? 

Also, would it make sense to actually fit the pipeline and check afterwards that it was fitted? Or maybe add that to some other unit test? As the issue was specifically about cloning, I think it would also make sense to fit, clone and then check if the estimator is still fitted, right?

franchuterivera replied:
Thank you, this makes a lot of sense and I added a push with this:

- Move the checking for unfitted components to the configuration test, this way we can also check post fit (as all the try/catch blocks are here)
- We have a cloning check, and it did not trigger the problem, because I mistakenly assumed self.models should be clonable objects, yet in the case of cross validation this attribute is not supposed to be used. I added a clarification note for future readers.

mfeurer commented on test/test_pipeline/test_regression.py:
```diff
@@ -586,3 +587,37 @@ def test_set_hyperparameters_honors_configuration(self):
+            for name, step in auto.named_steps.items():
+                with self.assertRaisesRegex(sklearn.exceptions.NotFittedError,
+                                            "instance is not fitted yet"):
+                    check_is_fitted(step)
```
Same as above.

franchuterivera replied:
Added this for regression as well!

mfeurer reviewed (commented):
Two minor comments on how I think the tests could be made stricter. Do you think that makes sense?

mfeurer commented on test/test_pipeline/test_regression.py:
```diff
@@ -169,8 +169,29 @@ def _test_configurations(self, configurations_space, make_sparse=False,
+                # After fit, the Regressor must be tagged as fitted
+                # by sklearn. Check is fitted raises an exception if that
+                # is not the case
+                # This is not enforced over every step of the pipeline, as
```
Would it make sense to enforce this to make this test stricter?

franchuterivera replied:
Yeah we can, adding something like:
https://github.com/scikit-learn/scikit-learn/blob/c3f25168fe8c73b82af8f57ae536f4a9708c863e/sklearn/metrics/_plot/tests/test_plot_precision_recall.py#L67

The only reason I did not do it is that it will require changing couple of places here and there, but if you think it makes sense I will go with the change.  Definitely it is a nice testing mechanism to ensure that the components were actually fitted. 

Would add a new push for this.

mfeurer replied:
> but if you think it makes sense I will go with the change

I'd appreciate that
